pkg/runner/local_docker.go

```diff
@@ -171,36 +171,82 @@ func (r *LocalDockerRunner) Healthcheck(ctx context.Context, engine api.Engine,
 	return hh.RunChecks(ctx, fix)
 }
 
-func (c *LocalDockerRunner) updateRunResult(template *runtime.RunParams, result *Result) error {
-	events, err := c.syncClient.FetchAllEvents(template)
+// setupSyncClient sets up the sync client if it is not set up already.
+func (r *LocalDockerRunner) setupSyncClient() error {
+	r.lk.Lock()
+	defer r.lk.Unlock()
+
+	if r.syncClient != nil {
+		return nil
+	}
+
+	// TODO: maybe make this a direct arg
+	err := os.Setenv(ss.EnvServiceHost, "127.0.0.1")
 	if err != nil {
 		return err
 	}
 
-	for _, e := range events {
-		// for now we emit only outcome OK events, so no need for more checks
-		if e.SuccessEvent != nil {
-			se := e.SuccessEvent
-			o := result.Outcomes[se.TestGroupID]
-			o.Ok = o.Ok + 1
-		}
+	r.syncClient, err = ss.NewGenericClient(context.Background(), logging.S())
+	if err != nil {
+		return err
 	}
 
-	result.Outcome = task.OutcomeSuccess
-	if len(result.Outcomes) == 0 {
-		result.Outcome = task.OutcomeFailure
+	return nil
+}
+
+func (r *LocalDockerRunner) collectOutcomes(ctx context.Context, result *Result, wg *sync.WaitGroup, tpl *runtime.RunParams) (context.CancelFunc, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	eventsCh, err := r.syncClient.SubscribeEvents(ctx, tpl)
+	if err != nil {
+		cancel()
+		return nil, err
 	}
-	for g := range result.Outcomes {
-		if result.Outcomes[g].Total != result.Outcomes[g].Ok {
+
+	wg.Add(1)
+	go func() {
+		running := true
+		for running {
+			select {
+			case <-ctx.Done():
+				running = false
+			case e := <-eventsCh:
+				// for now we emit only outcome OK events, so no need for more checks
+				if e.SuccessEvent != nil {
+					se := e.SuccessEvent
+					o := result.Outcomes[se.TestGroupID]
+					o.Ok = o.Ok + 1
+				}
+			}
+		}
+
+		result.Outcome = task.OutcomeSuccess
+		if len(result.Outcomes) == 0 {
 			result.Outcome = task.OutcomeFailure
-			break
 		}
-	}
 
-	return nil
+		for g := range result.Outcomes {
+			if result.Outcomes[g].Total != result.Outcomes[g].Ok {
+				result.Outcome = task.OutcomeFailure
+				break
+			}
+		}
+
+		wg.Done()
+	}()
+
+	return cancel, nil
 }
 
 func (r *LocalDockerRunner) Run(ctx context.Context, input *api.RunInput, ow *rpc.OutputWriter) (runoutput *api.RunOutput, err error) {
+	log := ow.With("runner", "local:docker", "run_id", input.RunID)
+
+	err = r.setupSyncClient()
+	if err != nil {
+		r.lk.Unlock()
```

Comment:
This is also a bit confusing. we are doing an `Unlock` and it is not at all obvious where the `Lock` happened?

Reply:
Shouldn't be there. Thanks for noticing!